contracts/utils/TokenVestingStaking.sol

```diff
@@ -6,259 +6,248 @@ import "../libraries/Shared.sol";
         IERC20 sent_
-    ) nzAddr(beneficiary_) nzAddr(address(stakingRewardsContract_)) nzAddr(address(sent_)) {
+    ) nzAddr(beneficiary_) nzAddr(address(rewardsContract_)) nzAddr(address(sent_)) {
         require(start_ <= end_, "Vesting: start_ after end_");
         require(block.timestamp < start_, "Vesting: start before current time");
 
-        beneficiary = beneficiary_;
-        revoker = revoker_;
-        start = start_;
-        end = end_;
+        beneficiary             = beneficiary_;
+        revoker                 = revoker_;
+        start                   = start_;
+        end                     = end_;
         transferableBeneficiary = transferableBeneficiary_;
-        stakingRewardsContract = stakingRewardsContract_;
-        SENT = sent_;
+        rewardsContract         = rewardsContract_;
+        snContribFactory        = snContribFactory_;
+        SENT                    = sent_;
     }
 
     //////////////////////////////////////////////////////////////
     //                                                          //
-    //                  State-changing functions                //
+    //                 Rewards contract functions               //
     //                                                          //
     //////////////////////////////////////////////////////////////
 
-    /**
-     * @notice Adds a BLS public key to the list of service nodes. Requires a proof of possession BLS signature to prove user controls the public key being added
-     * @param blsPubkey - 64 bytes of the bls public key
-     * @param blsSignature - 128 byte signature
-     * @param serviceNodeParams - Service node public key, signature proving ownership of public key and fee that operator is charging
-     */
     function addBLSPublicKey(
         BN256G1.G1Point calldata blsPubkey,
         IServiceNodeRewards.BLSSignatureParams calldata blsSignature,
-        IServiceNodeRewards.ServiceNodeParams calldata serviceNodeParams
-    ) external onlyBeneficiary notRevoked afterStart {
-        uint256 stakingRequirement = stakingRewardsContract.stakingRequirement();
-        uint64 serviceNodeID = stakingRewardsContract.nextServiceNodeID();
-        investorServiceNodes.push(ServiceNode(serviceNodeID, stakingRequirement));
-        SENT.approve(address(stakingRewardsContract), stakingRequirement);
-
-        // NOTE: Pass empty array, the contract will assume sender (this contract) as operator.
-        IServiceNodeRewards.Contributor[] memory contributors = new IServiceNodeRewards.Contributor[](0);
-        stakingRewardsContract.addBLSPublicKey(blsPubkey, blsSignature, serviceNodeParams, contributors);
-    }
+        IServiceNodeRewards.ServiceNodeParams calldata serviceNodeParams,
+        address addrToReceiveRewards
+    ) external onlyRevokerIfRevokedElseBeneficiary afterStart {
+        require(addrToReceiveRewards != address(0), "Rewards can not be paid to the zero-address");
```

Comment:
Perhaps we should also require `addrToReceiveRewards != address(this)` to prevent sending the (free) rewards back into the locked vesting contract.

Reply:
I have similar thoughts about to above about the what-if. There's one small advantage in that, if you as an investor want to reinvest everything back into the network, if you pool the rewards back into your contract you can fill in a bigger slot in a node if you had left-over $SENT in your contract that was too small to fill any existing nodes.

Again, maybe that's unfounded and the trade-off on accidentally locking your rewards is not worth it.

Reply:
Hmm yeah that might be useful somewhere.  I suppose we can just put the protection for this on the front-end or in docs or wherever.